recipe/2024-04-terraform-provider-secrets.md

```diff
@@ -0,0 +1,259 @@
+    }
+    recipeConfig: {
+      terraform: {
+        authentication: {
+          ...
+        }
+        providers: {
+          azurerm: [
+            {
+              subscriptionid: 1234
+              tenant_id: '745fg88bf-86f1-41af'
++              secrets: { // Individual Secrets from SecretStore
++                my_secret_1: {
++                  source: secretStoreConfig.id
++                  key: 'secret.one'
++                }
++                my_secret_2: {
++                  source: secretStoreConfig.id
++                  key: 'secret.two'
++                }
++              }
+            }, {
+              subscriptionid: 1234
+              tenant_id: '745fg88bf-86f1-41af'
+              alias: 'az-paymentservice'
+            } ]
+          gcp: [
+            {
+              project: 1234
+              regions: [ 'us-east1', 'us-west1' ]
+            }
+          ]
+          oraclepass: [
+            {
+              database_endpoint: '...'
+              java_endpoint: '...'
+              mysql_endpoint: '...'
+            }
+          ]
+        }
+      }
+      env: {
+        MY_ENV_VAR_1: 'my_value'
+      }
++      envSecrets: { // Individual Secrets from SecretStore
++        MY_ENV_VAR_2: {
++          source: secretStoreConfig.id
++          key: 'envsecret.one'
++        }
++        MY_ENV_VAR_3: {
++          source: secretStoreConfig.id
++          key: 'envsecret.two'
++        }
++     }
+    }
+    recipes: {
+    }
+  }
+}
+
+```
+
+**Sample Output:**
+n/a
+
+**Sample Recipe Contract:**
+n/a
+
+## Design
+
+### High Level Design
+
+The design for handling secrets for Private Terraform modules established the basis for data flow between Engine and Driver components in the codebase. For Terraform Providers, we will continue to maintain the flow where secrets are retrieved in the Engine and passed into the Driver in the Execute() call. 
+The following diagram describes the updated data flow. The calls in green highlight updates to existing code.
+
+
+![alt text](2024-04-terraform-provider-secrets.png)
+
+
+### Detailed Design
+
+The FindSecretIds function in the DriverWithSecrets interface is updated to return a struct of RecipeSecretIds, which will hold secret Ids for a Terraform module as before and in addition, also populate secret Ids for provider configuration and environment variables, iterating through the input environment recipe configuration. 
+
+The LoadSecrets method in the SecretLoader interface will also be updated to LoadRecipeSecrets and return struct RecipeSecrets which will contain secrets for module, providers and environment variables.
+
+#### Advantages (of each option considered)
+Retrieving secrets in the Engine and passing them into the Driver was discussed positively as a way forward. Looking ahead with Containerization, the Driver will have all the information it needs to be able to be spin off into a separate process.
+
+#### Disadvantages (of each option considered)
```

Comment:
There can be multiple secrets registered across different providers, but a particular recipe deployment might (most likely) won't require all of them. Is there a downside to passing all the extra information? Is there an alternative option?

Reply:
This will be discussed again based on input from PMs and their discussions with customers.

Reply:
> There can be multiple secrets registered across different providers, but a particular recipe deployment might (most likely) won't require all of them.

For my previous point above, can you explicitly call out this behavior in the doc?

> This will be discussed again based on input from PMs and their discussions with customers.

Let's write what specifically will we get from PMs and how that helps with the design, while the context of the discussion is still fresh in our mind. :)

Reply:
I had noted it under Design Review Notes, also added to Open Questions (we probably need it in only one section). I think I need more information about potential scenarios before writing more about it.

Reply:
If we are only going to have it in one section updating the source of truth is better than a notes section which I would consider to be the right place for meta observations vs changes.

Reply:
Updated the Detail Design section of document. Pls review.